### DISCUSSION TITLE
Strongswan threads always idle

### DISCUSSION THREAD

**Discussion by ptchuba** [strongswan/strongswan#1909] [Q&A]
I'm trying to see how many users I can have simultaneously connecting to my strongswan server within a short period of time. The responder is a gateway that is setup with my desired production configuration. As the initiator, I'm using strongswan with the load-tester plugin.  

I have tuned my server with info I found on the wiki and what I could glean from several previous posts on the topic. I can only get ~280 connections while trying to get 300. Trying more leads to more failures. But my CPU and Memory on the server and initiator seem barely touched. Server has 4 cores and 16GB RAM and i don't see the CPU going above 4% per core. Networking between the two boxes is great (both on azure and same region) and iperf3 shows almost 1Gbps between them.

I'm using 32 threads on the server as well as the initiator but have never seen the number of threads in use change on the `swanctl --stats` command. It's always showing 5 less than the number of threads specified as idle. And it's always 4/0/1/0 even though the job scheduled keeps going up and though I specified different number in the Job priority config.

Output from server:
```
root@Entry1:/home/acmeadmin# swanctl --stats
uptime: 2 minutes, since Sep 15 00:30:51 2023
worker threads: 32 total, 27 idle, working: 4/0/1/0
job queues: 0/0/0/0
jobs scheduled: 558
IKE_SAs: 279 total, 0 half-open
mallinfo: sbrk 11276288, mmap 0, used 10081600, free 1194688
loaded plugins: charon aesni aes rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm drbg attr kernel-netlink resolve socket-default connmark stroke vici updown eap-mschapv2 xauth-generic counters
root@Entry1:/home/acmeadmin#
```

 Here are my configs:

On Initiator
```
charon {
        load_modular = yes
        threads = 32
        plugins {
           load-tester {
              enable = yes
              # use 5 threads to initiate connections simultaneously
              initiators = 5
              # each thread initiates 60 connections
              iterations = 60
              # delay each initiation in each thread by 200ms
              delay = 500
              # fake the kernel interface to avoid SA conflicts
             # fake_kernel = yes
              responder = 201.62.174.276 #I know!
              initiator_auth = pubkey
              initiator_id = con2-%d-round-%d@acme.local
              initiator_match = *@acme.local
              #digest = sha384
              responder_id = entry1.acme.local
              request_virtual_ip = yes
              issuer_cert = /etc/swanctl/x509ca/rootCACert.pem
              issuer_key = /etc/swanctl/x509ca/rootCAKey.pem
              #esp = aes256gcm128-ecp384
              #proposal = aes256gcm16-prfsha384-ecp384
              # proposal = aes128-sha1-modpnull
              # disable IKE_SA rekeying (default)
              ike_rekey = 0
              # enable CHILD_SA every 60s
              child_rekey = 60
              # do not delete the IKE_SA after it has been established (default)
              delete_after_established = no
              # do not shut down the daemon if all IKE_SAs established
              shutdown_when_complete = no

              }
           include strongswan.d/charon/*.conf
        }
}
```

strongswan.conf on server:
```
charon {
        load_modular = yes
        threads = 32
        keep_alive = 0
        ikesa_table_size = 512
        ikesa_table_segments = 32
        reuse_ikesa = no
        dos_protection = no
        retransmit_timeout = 10
        retransmit_base = 2
        half_open_timeout = 90
        init_limit_half_open = 500
        plugins {
                include strongswan.d/charon/*.conf
        }
        processor {
          priority_threads {
            high = 2
            medium = 8
          }
        }
}
```

swanctl on server
```
   clients {
      local_addrs = %any
      pools = client_pool_v4
      #, client_pool_v6
      send_cert = always
      local {
         auth = pubkey
         certs = entry1-cert.pem
         id = entry1.acme.local
      }
      remote {
         auth = pubkey
      }
      children {
         client-sa {
            local_ts = 0.0.0.0/0,::/0
            if_id_out = 2
            if_id_in = 2
            esp_proposals = aes128-sha1,aes256gcm128-ecp384
            mode = tunnel
         }
      }
      version = 2
      encap = yes
      proposals = aes128-sha1-modp768
      #aes256gcm16-prfsha512-ecp384,aes256gcm16-prfsha384-ecp384
   }
```

I'm suspecting the issue is with the specified number of threads not being used. 
Any help will be greatly appreciated!!

**Comment by tobiasbrunner** [comment]
> I can only get ~280 connections while trying to get 300. Trying more leads to more failures.

What kind of failures?

>               # use 5 threads to initiate connections simultaneously
>               initiators = 5
>               # each thread initiates 60 connections
>               iterations = 60
>               # delay each initiation in each thread by 200ms
>               delay = 500

This gives you 10 initiations per second (5 initiators each waiting 0.5s between initiations). That's not very much, so it's not really surprising that the server is basically idle.
